pkg/apis/core/validation/validation.go

```diff
@@ -6974,3 +7047,264 @@ func sameLoadBalancerClass(oldService, service *core.Service) bool {
 	}
 	return *oldService.Spec.LoadBalancerClass == *service.Spec.LoadBalancerClass
 }
+
+// ValidateResourceClaim validates a ResourceClaim.
+func ValidateResourceClaim(resourceClaim *core.ResourceClaim) field.ErrorList {
+	allErrs := ValidateObjectMeta(&resourceClaim.ObjectMeta, true, validateResourceClaimName, field.NewPath("metadata"))
+	allErrs = append(allErrs, validateResourceClaimSpec(&resourceClaim.Spec, field.NewPath("spec"))...)
+	return allErrs
+}
+
+var validateResourceClaimName = apimachineryvalidation.NameIsDNSSubdomain
+
+func validateResourceClaimSpec(spec *core.ResourceClaimSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for _, msg := range ValidateClassName(spec.ResourceClassName, false) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("resourceClassName"), spec.ResourceClassName, msg))
+	}
+	allErrs = append(allErrs, validateResourceClaimParameters(spec.ParametersRef, fldPath.Child("parametersRef"))...)
+	switch spec.AllocationMode {
+	case core.AllocationModeImmediate, core.AllocationModeWaitForFirstConsumer:
+	default:
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("allocationMode"), spec.AllocationMode, "unknown mode"))
+	}
+	return allErrs
+}
+
+// It would have been nice to use Go generics to reuse the same validation
+// function for Kind and Name in both types, but generics cannot be used to
+// access common fields in structs.
+
+func validateResourceClaimParameters(ref *core.ResourceClaimParametersReference, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if ref != nil {
+		if ref.Kind == "" {
+			allErrs = append(allErrs, field.Required(fldPath.Child("kind"), "must be specified"))
+		}
```

Comment:
It's a reference to arbitrary other objects, therefore it is not known what a well-formed name for those looks like. It's also not possible to check that kind is known because that may change over time (think CRDs).